src/use-transition/use-transition.ts

```diff
@@ -0,0 +1,154 @@
+  const { open, hiddenClass } = Object.assign(defaultOptions, options)
+  const targetElement = options?.element || controller.element
+  if (!((targetElement instanceof HTMLElement) || (targetElement instanceof SVGElement))) return
+  const dataset = targetElement.dataset
+
+  const controllerShow: Function = controller.show?.bind(controller)
+  const controllerHide: Function = controller.hide?.bind(controller)
+  const controllerToggle: Function = controller.toggle?.bind(controller)
+
+  async function show(event?: Event) {
+    if (controller.isOpen) return
+
+    controller.isOpen = true
+    controllerShow && controllerShow(event)
+
+    const enterClass = getAttribute("enter", options, dataset)
+    const enterActiveClass = getAttribute("enterActive", options, dataset)
+    const enterToClass = getAttribute("enterTo", options, dataset)
+
+    if (!!hiddenClass) {
+      targetElement.classList.remove(hiddenClass)
+    }
+
+    await transition(targetElement, enterClass, enterActiveClass, enterToClass)
+
+  }
+
+  async function hide(event?: Event) {
+    if (!controller.isOpen) return
+
+    controller.isOpen = false
+    controllerHide && controllerHide(event)
+
+    const leaveClass = getAttribute("leave", options, dataset)
+    const leaveActiveClass = getAttribute("leaveActive", options, dataset)
+    const leaveToClass = getAttribute("leaveTo", options, dataset)
+
+    await transition(targetElement, leaveClass, leaveActiveClass, leaveToClass)
+
+    if (!!hiddenClass) {
+      targetElement.classList.add(hiddenClass)
+    }
+  }
+
+  function toggle(event: Event) {
+    controllerToggle && controllerToggle(event)
+    if (controller.isOpen) {
+      hide()
+    } else {
+      show()
+    }
+  }
+
+  async function transition(element: Element, initialClass: string[], activeClass: string[], toClass: string[]) {
+    element.classList.add(...initialClass);
+    element.classList.add(...activeClass);
+    await nextFrame();
+
+    element.classList.remove(...activeClass);
+    element.classList.add(...toClass);
+
+    await afterTransition(element);
+
+    element.classList.remove(...initialClass);
+  }
+
+  function initialState() {
+    if (open) {
+      if (!!hiddenClass) {
+        targetElement.classList.remove(hiddenClass)
+      }
+      show()
+    } else {
+      if (!!hiddenClass) {
+        targetElement.classList.add(hiddenClass)
+      }
+      hide()
+    }
+  }
+
+  initialState()
+  Object.assign(controller, { show, hide, toggle })
+}
+
+function getAttribute(name: string, options: TransitionOptions, dataset: DOMStringMap) {
+  const datasetName = `transition${name[0].toUpperCase()}${name.substr(1)}`
+  const datasetAlpineName = (alpineNames as any)[name]
+  const classes = (options as any)[name] || dataset[datasetName] || dataset[datasetAlpineName] || ""
+  return classes.split(" ")
+}
+
+async function afterTransition(element: Element): Promise<Function> {
```

Comment:
I've implemented something similar before. Here's my implementation for `afterTransition` that handles multiple different transition durations on an element:
```
function singleTimeValueToMs(duration: string): number {
    const matches = duration.match(/^(-?[0-9.]+)(s|ms)$/);

    const invalidValueError = new Error('Invalid time value');

    if (matches === null) {
        throw invalidValueError;
    }

    if (matches.length !== 3) {
        throw invalidValueError;
    }

    const [match, length, unit] = matches;

    if (unit === 's') {
        return Number(length) * 1000;
    }

    return Number(length);
}

function multiTimeValueToMs(durations: string): number[] {
    return durations.split(',')
        .map(str => str.trim())
        .map(singleTimeValueToMs);
}

function afterTransition(element) {
    return new Promise(resolve => {
        const durations = multiTimeValueToMs(
            window.getComputedStyle(element).transitionDuration
        );

        setTimeout(resolve, Math.max(...durations));
    });
}
```

Reply:
Thanks @panda-madness will definitely check it out